### PR TITLE
initial commit of the newest pdf scraper

### PR DESCRIPTION
**Pull request by DanaeBouwer**
**Major Adjustments and restructuring of pdf scraper**

_Summary_
The new scraper takes its input from the masterlist, and loops through a list of article ID's. It directly navigates to the article download page on JSTOR, and automatically downloads it from there. This scraper also integrates the reCAPTCHA solver which was restructured by incorporating cookies to make sure the user stays logged in for the mp3 download and audio recognition to text transformation.

### CODE REVIEW

File: Masterlist_Scraper.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
+    chrome_options.add_experimental_option("useAutomationExtension", False)
+    chrome_options.add_experimental_option("prefs", {
+    "download.default_directory": directory, #Change default directory for downloads
+    "download.prompt_for_download": False, #To auto download the file
+    "download.directory_upgrade": True,
+    "plugins.always_open_pdf_externally": True, #It will not show PDF directly in chrome
+    "credentials_enable_service": False, # gets rid of password saver popup
+    "profile.password_manager_enabled": False #gets rid of password saver popup
+    })
```
Comment (wenkanglu) [inline Masterlist_Scraper.py]: This can all be outside of the loop since none of it changes between iterations.

File: Masterlist_Scraper.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    cookies = pickle.load(open("cookies.pkl", "rb"))
+    for cookie in cookies:
+        driver.add_cookie(cookie)
+
+    # else:
+    #     web_session = driver.get("https://www.jstor.org")
+
+    #     cookies = pickle.load(open("cookies.pkl", "rb"))
+    #     for cookie in cookies:
+    #         driver.add_cookie(cookie)
```
Comment (wenkanglu) [inline Masterlist_Scraper.py]: Remove if not needed.

**Review by wenkanglu — CHANGES_REQUESTED** [review]
General comments:
* We should use a dependency manager like [_Poetry_](https://python-poetry.org/) to keep track of dependencies of the project. It'll also allow whoever else working on this repository to easily download all dependencies.
* Make use of a formatter like [_Black_](https://github.com/psf/black) for consistent styling.
* Follow PEP8 for additional styling and naming conventions.
* Rename Masterlist_Scraper.py to main.py since it is the entry point to the application.
* Remove unnecessary committed files/folders like __pycache__.
* Remove Metadata.csv and Metadata.json and store them in something like Google Drive and point them out in a README. The files are too large and makes pushing and pulling cumbersome.

File: Masterlist_Scraper.py (hunk 1 above)
Comment (DanaeBouwer) [reply Masterlist_Scraper.py]: Cool, done!

File: Masterlist_Scraper.py (hunk 2 above)
Comment (DanaeBouwer) [reply Masterlist_Scraper.py]: Made the change.

File: src/connection_controllers/gen_connection_controller.py [hunk 3]
```diff
@@ -5,31 +5,30 @@
 from selenium.webdriver.common.by import By
 from selenium import webdriver
 
-class GenConnectionController(ConnectionController):
 
+class GenConnectionController(ConnectionController):
     def __init__(self, driver: webdriver, host: str):
```
Comment (wenkanglu) [inline src/connection_controllers/gen_connection_controller.py]: I do not recommend placing any logic in the __init__ function. The __init__ function serves as the "constructor" to our class and should only be used to assign local variables to self. Any additional logic should be placed in their own function and then called upon externally.

File: src/main.py [hunk 4]
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
...
 
-        # Construct file names
-        url = os.path.join(directory,article.split("/")[-1] + ".pdf")
-        url_pending = os.path.join(directory,article.split("/")[-1] + ".pdf.crdownload")
-        doi = os.path.join(directory,"10.2307." + article.split("/")[-1] + ".pdf")
-        
         # The t&c's only pop up the first time you start the driver
 
-        while t_c_accepted!=True:
+        while t_c_accepted != True:
```
Comment (wenkanglu) [inline src/main.py]: Can be simplified to `while !t_c_accepted:`

Also, this is potentially infinite if something goes wrong and the t&c's are never accepted. We should add a maximum number of tries (same as any other potentially infinite while loop).

You can do something like `while !t_c_accepted and current_try < MAX_TRIES_FOR_T&C`

In fact, once we do this, we can then log when this maximum number of tries has taken place and we can see where our bottlenecks are during analysis.

File: src/main.py [hunk 5]
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
...
+                if recaptcha_solver(driver, url, url_pending) == True:
                     print("solved")
                     continue
                 else:
                     print("[ERR] reCAPTCHA could not be solved, restart driver")
                     restart = True
                     break
-        
+
         if restart == True:
```
Comment (wenkanglu) [inline src/main.py]: `if restart`

File: src/main.py [hunk 6]
```diff
@@ -26,156 +27,172 @@ def latest_download_file():
...
             else:
                 print("[ERR] reCAPTCHA could not be solved, restarting driver")
                 break
 
         # check if file has downloaded - add a max time limit?
         file = url_pending
-        #count = 0
-        while file == url_pending: #and count <= 60:
+        # count = 0
+        while file == url_pending:  # and count <= 60:
```
Comment (wenkanglu) [inline src/main.py]: Why is the count commented out? We want to avoid infinite loops at all cost.

File: src/recaptcha_solver.py [hunk 7]
```diff
@@ -12,17 +12,23 @@
 import pickle
 import requests
 
+
 def delay():
```
Comment (wenkanglu) [inline src/recaptcha_solver.py]: Rather add the minimum and maximum time to wait as parameters so you can reuse this delay function wherever you need a random amount of time to wait.

File: src/recaptcha_solver.py [hunk 8]
```diff
@@ -44,44 +52,47 @@ def check_solved(driver,url,url_pending):
...
 
-def recaptcha_solver(driver,url,url_pending):
+
+def recaptcha_solver(driver, url, url_pending):
 
     recaptcha_log = 0
 
     recaptcha_control = True
 
     while recaptcha_control == True:
```
Comment (wenkanglu) [inline src/recaptcha_solver.py]: `while recaptcha_control`
Although the variable name does not really sound like a boolean. Maybe think of a different name here.

File: src/recaptcha_solver.py [hunk 9]
```diff
@@ -94,56 +105,62 @@ def recaptcha_solver(driver,url,url_pending):
...
             break
 
-        if check_solved(driver,url,url_pending)==True:
+        if check_solved(driver, url, url_pending) == True:
             success = True
             break
         else:
             recaptcha_challenge = True
-            while recaptcha_challenge==True:
+            while recaptcha_challenge == True:
```
Comment (wenkanglu) [inline src/recaptcha_solver.py]: `while recaptcha_challenge`

Change name to be more of a boolean.

File: src/recaptcha_solver.py [hunk 10]
```diff
@@ -213,7 +236,7 @@ def recaptcha_solver(driver,url,url_pending):
                         recaptcha_control = False
                         break
 
-                    if check_solved(driver,url,url_pending)==True:
+                    if check_solved(driver, url, url_pending) == True:
```
Comment (wenkanglu) [inline src/recaptcha_solver.py]: ` if check_solved(driver, url, url_pending):`

File: src/temp_storage.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import os
+from pathlib import Path
+
+
+class storage:
+    def getTempStoragePath():
```
Comment (wenkanglu) [inline src/temp_storage.py]: Please be consistent with naming. In Python we prefer snake case, e.g., `get_temp_storage_path`

**Review by wenkanglu — CHANGES_REQUESTED** [review]
Rather minor changes that you can just add with your next commit.

File: src/main.py (hunk 4 above)
Comment (DanaeBouwer) [reply src/main.py]: The while !t_c_accepted is not working, but adjusted that section to avoid a possible infinite loop.

File: src/main.py (hunk 6 above)
Comment (DanaeBouwer) [reply src/main.py]: It was for a test, have uncommented it.

File: src/recaptcha_solver.py (hunk 7 above)
Comment (DanaeBouwer) [reply src/recaptcha_solver.py]: I added a module that calculates the 5 point moving average download time of the user. I replaced the delay function with this and re-use it throughout the programme.

File: src/recaptcha_solver.py (hunk 8 above)
Comment (DanaeBouwer) [reply src/recaptcha_solver.py]: There are two frames in the reCAPTCHA: a control frame and a challenge frame. I therefore called the respective while loops for each section, recaptcha_control and recaptcha_challenge. Do you still thing the naming convention should be changed?

File: src/main.py (hunk 4 above)
Comment (wenkanglu) [reply src/main.py]: Try "while not t_c_accepted". Forgot that this was Python lol

File: src/recaptcha_solver.py (hunk 9 above)
Comment (DanaeBouwer) [reply src/recaptcha_solver.py]: Same as the recaptcha_control. Let me know if I should change it.

File: src/temp_storage.py (hunk 11 above)
Comment (DanaeBouwer) [reply src/temp_storage.py]: Got it, thanks!

File: src/recaptcha_solver.py (hunk 8 above)
Comment (wenkanglu) [reply src/recaptcha_solver.py]: Yeah so a boolean should always be asking a question in the form of "Is something true or false". A recaptcha control or challenge being true is vague as to what that really means. For example, if it is regarding whether the control is active, then `is_recaptcha_control_active` is a better name.

Comment (DanaeBouwer) [reply src/recaptcha_solver.py]: Okay, makes sense! Will change the naming appropriately.

**Pull request merged**